app/features/settings/profile/editable-username.tsx

```diff
@@ -0,0 +1,153 @@
+import { Check, Edit } from 'lucide-react';
+import { useEffect, useRef, useState } from 'react';
+import { useForm } from 'react-hook-form';
+import { Button } from '~/components/ui/button';
+import { useUpdateUsername, useUser } from '~/features/user/user-hooks';
+import useAnimation from '~/hooks/use-animation';
+import { useToast } from '~/hooks/use-toast';
+import { cn } from '~/lib/utils';
+
+// Reserved routes that can't be used as usernames
+// QUESTION: can we do this more efficiently so that
+// we don't have to update for every new route?
+const RESERVED_ROUTES = [
+  'settings',
+  'profile',
+  'login',
+  'signup',
+  'api',
+  'admin',
+  'help',
+  'about',
+  'terms',
+  'privacy',
+  'receive',
+  'send',
+  'verify-email',
+];
+
+type FormValues = {
+  username: string;
+};
+
+function validateUsername(username: string) {
+  // Check length (3-20 characters)
+  if (username.length < 3 || username.length > 20) {
+    return 'Username must be between 3 and 20 characters';
+  }
+
+  // Check characters (only a-z, 0-9, underscore, hyphen allowed)
+  // This requirement comes from LUD16 https://github.com/lnurl/luds/blob/luds/16.md
+  if (!/^[a-z0-9_-]+$/.test(username)) {
+    return 'Username can only contain lowercase letters, numbers, underscores and hyphens';
+  }
+
+  if (RESERVED_ROUTES.includes(username.toLowerCase())) {
+    return 'This username is not available';
+  }
+
+  return true;
+}
+
+export default function EditableUsername() {
+  const { animationClass, start: startAnimation } = useAnimation({
+    name: 'slam',
+    durationMs: 400,
+  });
+  const { toast } = useToast();
+  const inputRef = useRef<HTMLInputElement>(null);
+  const user = useUser();
+  const updateUsername = useUpdateUsername();
+  const {
+    register,
+    handleSubmit,
+    watch,
+    formState: { errors },
+  } = useForm<FormValues>({
+    defaultValues: {
+      username: user.username,
+    },
+  });
+  const [isEditing, setIsEditing] = useState(false);
+
+  useEffect(() => {
+    isEditing && inputRef.current?.focus();
+  }, [isEditing]);
+
+  const onSubmit = async (data: FormValues) => {
+    if (data.username === user.username) {
+      setIsEditing(false);
+      return;
+    }
+
+    try {
+      await updateUsername(data.username);
+      startAnimation();
+      setIsEditing(false);
+    } catch {
+      toast({
+        title: 'Error',
+        description: 'Failed to update username',
+      });
+    }
+  };
+
+  const handleEditClick = (e: React.MouseEvent) => {
+    e.preventDefault();
+    setIsEditing(true);
+  };
+
+  return (
+    <form
+      onSubmit={handleSubmit(onSubmit)}
+      className={'flex w-full flex-col gap-2'}
+    >
+      <div className="flex items-center gap-2">
+        <div className={cn('flex-1', animationClass)}>
+          {isEditing ? (
+            <input
+              {...register('username')}
+              ref={(e) => {
```

Comment:
what is this ref for? can't we register field and validation the same way we did in our other forms?